### PR TITLE
Attempt to define resolution and dereferencing

### PR DESCRIPTION
**Pull request by OR13**
This PR attempts to address the problem of documenting the various ways that known key content types, such as `application/jwk+json` and `application/cose-key` are discovered from opaque strings or URIs such as `iss` and `kid`.

This PR also explains how DIDs are similar to OIDC.

### CODE REVIEW

File: draft-ietf-scitt-architecture.md [hunk 1]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+      "crv": "P-384",
+      "alg": "ES384",
+      "x": "LCeAt2sW36j94wuFP0gNEIHDzqR6Nh_Udu2ObLer3cKFBCaAHY1svmbPV69bP3RH",
+      "y": "zz2SkcOGYM6PbYlw19tcbpzo6bEMYHIwGBnN5rd8QWykAprstPdxx4U0uScvDcYd"
+    }
+  }]
+}
+```
+
+Editor note, we might wish to eliminate this intermediate identity document content type,
```
Comment (OR13) [inline draft-ietf-scitt-architecture.md:430]: @Sakurann this is related to the conversations we have had on https://www.w3.org/TR/did-core/#verification-relationships

... if you use those JSON-LD relationships would cannot eliminate did documents, or you need to encode the relationship in the identifier itself... for example:

```
did:example:123/assertion/kid => jwk+json
did:example:123/assertion=> jwk-set+json
```

File: draft-ietf-scitt-architecture.md [hunk 2]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+by treating it as an alterative encoding of `application/jwk-set+json` or `application/cose-key-set`.
+
+However, there is no media type fragment processing directive
+that would enable dereferencing the known key set content types, listed above.
+
+##### Comment on OIDC
```
Comment (OR13) [inline draft-ietf-scitt-architecture.md:436]: See a similar section of the W3C draft: https://w3c.github.io/vc-jwt/#key-discovery

It's a mess, because of mistakes made in DID Core... So we expect anything taking a dependency on DIDs to have similar sections of their spec.

**Comment by OR13** [comment]
I have a working demo of this on scitt.xyz, 

`https://transmute.id/did:web:scitt.xyz`

`https://scitt.xyz/.well-known/did.json`

Its also possible to do same thing with did methods other than `did:web`, for example `did:jwk`...

There is a demo of that working with index db here: https://scitt.xyz/demo

This shows that there is interoperability across issuers that use different did methods, that conform to the same resolution and dereferencing interfaces.

**Review by henkbirkholz — APPROVED** [review]
a proposal for discussion @ IETF 117 Hackathon and meeting seems very useful to me.

:+1:

File: draft-ietf-scitt-architecture.md [hunk 3]
```diff
@@ -371,6 +371,102 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
 They MAY update their DID manifest, for instance to refresh their signing keys or algorithms, but they SHOULD NOT remove or change any prior keys unless they intend to revoke all Signed Statements that are registered as Transparent Statements issued with those keys.
 This DID appears in the Issuer protected header of Signed Statements' Envelopes, while the version of the key from the manifest used to sign the Signed Statement is written in the `kid` header.
 
+`kid` MUST either be an absolute URL, 
+or `iss` MUST also be present in the protected header, 
+and `kid` MUST be a URL relative to `iss`.
```
Comment (roywill) [inline draft-ietf-scitt-architecture.md:376]: How about:
and 'kid' MUST then be a URL relative to `iss`

**Review by roywill — APPROVED** [review]
Asked Kristina to also look at this.

File: draft-ietf-scitt-architecture.md [hunk 4]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
 They MAY update their DID manifest, for instance to refresh their signing keys or algorithms, but they SHOULD NOT remove or change any prior keys unless they intend to revoke all Signed Statements that are registered as Transparent Statements issued with those keys.
 This DID appears in the Issuer protected header of Signed Statements' Envelopes, while the version of the key from the manifest used to sign the Signed Statement is written in the `kid` header.
 
+`kid` MUST either be an absolute URL,
+or `iss` MUST also be present in the protected header,
+and `kid` MUST be a URL relative to `iss`.
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:376]: Since `iss` seems to be required to be a DID ([this reference?](https://datatracker.ietf.org/doc/html/draft-ietf-scitt-architecture-01#section-6.1-4.2)), would it be clearer to say absolute/relative DID URL?

Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:376]:
```suggestion
`kid` MUST either be an absolute URL,
or a relative URL. Relative URL MUST be
relative to an `iss` value. When relative URL is used,
`iss` MUST also be present in the protected header.
```

I think the requirement conveyed can be clearer.

Comment (Sakurann) [reply draft-ietf-scitt-architecture.md:376]: not sure if it should say absolute DID URL and relative DID URL which are terms used in [DID Core](https://www.w3.org/TR/did-core/#relative-did-urls). - depends on the response to a question above.

File: draft-ietf-scitt-architecture.md [hunk 5]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+
+Resolving `kid` MUST return an Identity Document of a registered content type (a set of public keys).
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:378]:
```suggestion
Resolving `kid` MUST return an identity document of a registered content type (a set of public keys).
```
let's be consistent in capitalization.

File: draft-ietf-scitt-architecture.md [hunk 6]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+In the case of `kid` being an absolute DID URL, the identity document is called a DID Document,
+and is expected ot have content type `application/did+json`.
+
+Dereferencing a DID URL requires that it first be resolved, and then the fragment processed according to the media type.
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:382]:
```suggestion
To dereference a DID URL, it first MUST be resolved. After that the fragment is processed according to the media type.
```

File: draft-ietf-scitt-architecture.md [hunk 7]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+
+For example, when resolving `did:example:123#key-42`,
+first the identity document for `did:example:123` is resolved as content type `application/did+json`,
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:385]:
```suggestion
first, the identity document for `did:example:123` is resolved as content type `application/did+json`,
```

File: draft-ietf-scitt-architecture.md [hunk 8]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+next, the fragment `#key-2` is dereferenced to a verification method that contains a `publicKeyJwk` property in JSON.
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:386]: `in JSON` here should be removed or made clearer.

File: draft-ietf-scitt-architecture.md [hunk 9]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+
+The content type of `publicKeyJwk` is expected to be `application/jwk+json`.
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:388]: why is there a need for a content type for a JWK? JWK is a published RFC and has a clearly defined content/structure

File: draft-ietf-scitt-architecture.md [hunk 10]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+
+The details of both `did resolution` and `did dereferencing` are out of scope for this document.
+
+The `iss` or `kid`, might not be DID URLs, however the following interfaces MUST be satisfied in order to ensure
+issuer identity documents, and associated keys are discoverable in a consistent manner.
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:393]: I think this should be moved before line 379.

File: draft-ietf-scitt-architecture.md [hunk 11]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+
+Dereferencing a DID URL requires that it first be resolved, and then the fragment processed according to the media type.
+
+For example, when resolving `did:example:123#key-42`,
+first the identity document for `did:example:123` is resolved as content type `application/did+json`,
+next, the fragment `#key-2` is dereferenced to a verification method that contains a `publicKeyJwk` property in JSON.
+
+The content type of `publicKeyJwk` is expected to be `application/jwk+json`.
+
+The details of both `did resolution` and `did dereferencing` are out of scope for this document.
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:390]:
```suggestion
The details of both `DID resolution` and `DID dereferencing` are out of scope for this document.
```

File: draft-ietf-scitt-architecture.md [hunk 12]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
 They MAY update their DID manifest, for instance to refresh their signing keys or algorithms, but they SHOULD NOT remove or change any prior keys unless they intend to revoke all Signed Statements that are registered as Transparent Statements issued with those keys.
 This DID appears in the Issuer protected header of Signed Statements' Envelopes, while the version of the key from the manifest used to sign the Signed Statement is written in the `kid` header.
 
+`kid` MUST either be an absolute URL,
+or `iss` MUST also be present in the protected header,
+and `kid` MUST be a URL relative to `iss`.
+
+Resolving `kid` MUST return an Identity Document of a registered content type (a set of public keys).
+In the case of `kid` being an absolute DID URL, the identity document is called a DID Document,
+and is expected ot have content type `application/did+json`.
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:380]: what about relative DID URLs??

also a reference to the definition of relative DID URL in [DID Core](https://www.w3.org/TR/did-core/#relative-did-urls) might be helpful?

File: draft-ietf-scitt-architecture.md [hunk 13]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+next, the fragment `#key-2` is dereferenced to a verification method that contains a `publicKeyJwk` property in JSON.
+
+The content type of `publicKeyJwk` is expected to be `application/jwk+json`.
+
+The details of both `did resolution` and `did dereferencing` are out of scope for this document.
+
+The `iss` or `kid`, might not be DID URLs, however the following interfaces MUST be satisfied in order to ensure
+issuer identity documents, and associated keys are discoverable in a consistent manner.
+
+#### Resolving Identity Documents
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:395]: how is this section related to a sentence above `The details of both `did resolution` and `did dereferencing` are out of scope for this document.` should be clearer

File: draft-ietf-scitt-architecture.md [hunk 14]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+by treating it as an alterative encoding of `application/jwk-set+json` or `application/cose-key-set`.
+
+However, there is no media type fragment processing directive
+that would enable dereferencing the known key set content types, listed above.
+
+##### Comment on OIDC
+
+For well known token types, such as `id_token` or `access_token`.
+
+`iss` MUST be a URL, and it MUST have keys discoverable in the following way:
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:440]: I don't think there is a need to copy what is already defined in OIDC.Core and rfc6749. just pointing to the related sections in those specs is sufficient.

File: draft-ietf-scitt-architecture.md [hunk 15]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+      ]
+    }
+  ]
+}
+
+```
+
+If SCITT wanted to be interoperable with OIDC, we would define key dereferencing in a way that was compatible with how OIDC handles it today.
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:487]: I am still catching up, but using ID Tokens/Access Tokens here intuitively seems odd. Can Issuer be a natural person authenticated by OIDC?

File: draft-ietf-scitt-architecture.md (hunk 2 above)
Comment (Sakurann) [reply draft-ietf-scitt-architecture.md:436]: I think `.well-known/jwt-issuer` defined in SD-JWT VC draft is more relevant here. https://vcstuff.github.io/draft-terbu-sd-jwt-vc/draft-terbu-oauth-sd-jwt-vc.html#name-jwt-issuer-metadata

File: draft-ietf-scitt-architecture.md [hunk 16]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+`id` = `kid` if `iss` is undefined, or `iss` + `#` + `kid` when `iss` is defined.
+
+See also [draft-ietf-cose-cwt-claims-in-headers](https://datatracker.ietf.org/doc/draft-ietf-cose-cwt-claims-in-headers/).
+
+```
+dereference = (id: string, accept: content_type = 'application/jwk+json') =>
+publicKeyJwk (of content type application/jwk+json).
+```
+
+For example:
```
Comment (Sakurann) [inline draft-ietf-scitt-architecture.md:506]:
```suggestion
For example, when DIDs are used:
```

**Review by Sakurann — CHANGES_REQUESTED** [review]
made comments, but I think a bit more discussion/work is needed.

File: draft-ietf-scitt-architecture.md [hunk 17]
```diff
@@ -371,6 +371,156 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
...
+
+Resolving `kid` MUST return an identity document of a registered content type (a set of public keys).
+In the case of `kid` being an absolute DID URL, the identity document is called a DID Document,
+and is expected ot have content type `application/did+json`.
+
+To dereference a DID URL, it first MUST be resolved. After that the fragment is processed according to the media type.
+
+For example, when resolving `did:example:123#key-42`,
+first, the identity document for `did:example:123` is resolved as content type `application/did+json`,
+next, the fragment `#key-2` is dereferenced to a verification method that contains a `publicKeyJwk` property in JSON.
```
Comment (OR13) [inline draft-ietf-scitt-architecture.md:386]:
```suggestion
next, the fragment `#key-2` is dereferenced to a verification method that contains a `publicKeyJwk`.
```

File: draft-ietf-scitt-architecture.md (hunk 12 above)
Comment (OR13) [reply draft-ietf-scitt-architecture.md:381]: I agree, I filed an issue for this: https://github.com/ietf-wg-scitt/draft-ietf-scitt-architecture/issues/81

File: draft-ietf-scitt-architecture.md (hunk 9 above)
Comment (OR13) [reply draft-ietf-scitt-architecture.md:389]: Just being explicit.. and... because resolution and dereferencing are only described in the abstract in did core.

Most of this text would not be needed if DID Core defined resolution and dereferencing more concretely

File: draft-ietf-scitt-architecture.md (hunk 10 above)
Comment (OR13) [reply draft-ietf-scitt-architecture.md:394]: It's definitely a separate issue, the list is currently discussing all the possible ways to identify an organization with public keys... we are pretty sure we want to support DIDs, but we don't know how many other schemes we should support.

File: draft-ietf-scitt-architecture.md (hunk 13 above)
Comment (OR13) [reply draft-ietf-scitt-architecture.md:396]: Hard to fix, given DID Core doesn't define them either.

Any more specific suggestion here?

File: draft-ietf-scitt-architecture.md (hunk 2 above)
Comment (OR13) [reply draft-ietf-scitt-architecture.md:437]: fwiw, I plan to add that to the vc-jwt draft when it lands as well...

However, I disagree that drafts of "VCs" are more relevant than OIDC, which has massive adoption.

File: draft-ietf-scitt-architecture.md (hunk 15 above)
Comment (OR13) [reply draft-ietf-scitt-architecture.md:488]: It's an example... the point is JWTs are useless until they are profiled in a way that easier to adopt and ensures interop... thats what OIDC did for JWT, what will SCITT do for signed statements about software artifacts? 

This is probably not conveyed well, and possibly more directly to the working group than readers of the draft... I suspect this whole section will need several iterations to get right.

File: draft-ietf-scitt-architecture.md [hunk 18]
```diff
@@ -371,6 +371,157 @@ Issuers SHOULD use consistent decentralized identifiers for all their Statements
 They MAY update their DID manifest, for instance to refresh their signing keys or algorithms, but they SHOULD NOT remove or change any prior keys unless they intend to revoke all Signed Statements that are registered as Transparent Statements issued with those keys.
 This DID appears in the Issuer protected header of Signed Statements' Envelopes, while the version of the key from the manifest used to sign the Signed Statement is written in the `kid` header.
 
+`kid` MUST either be an absolute URL,
+or a relative URL. Relative URL MUST be
+relative to an `iss` value. When relative URL is used,
+`iss` MUST also be present in the protected header.
+
+Resolving `kid` MUST return an identity document of a registered content type (a set of public keys).
+In the case of `kid` being an absolute DID URL, the identity document is called a DID Document,
```
Comment (SteveLasker) [inline draft-ietf-scitt-architecture.md:380]: nit: linebreak on sentences across the PR

**Review by SteveLasker — APPROVED** [review]
LGTM to merge as a baseline for future discussion

File: draft-ietf-scitt-architecture.md (hunk 8 above)
Comment (OR13) [reply draft-ietf-scitt-architecture.md]: agredd, I removed it

File: draft-ietf-scitt-architecture.md (hunk 18 above)
Comment (OR13) [reply draft-ietf-scitt-architecture.md:380]: fix it editorial
